### DISCUSSION TITLE
Harmonizing data explorer and live dashboard

### DISCUSSION THREAD

**Discussion by dominikriemer** [apache/streampipes#3007] [Dev]
Hi,
currently, we provide two views to visualize data. 
The live dashboard aims at providing a dashboard which provides real-time data. The data explorer aims at providing a way to visually analyze historical data. Both views have similar concepts, e.g., users can create dashboards and place widgets on dashboards.
While I think this generally makes sense since the use cases (real-time monitoring vs. visual analytics) are different, both components don't share the same concepts and provide very different charts and chart settings. Maintaining both views is rather hard and we haven't really managed to improve the live dashboard in the last years.

I'd like to discuss harmonization of the dashboard and data explorer as follows:

* We keep both top-level views `Dashboard` and `Data Explorer`.
* In the data explorer view, users can create and view charts. Charts can provide their own time settings and users can create or edit a single chart at any time.
* In the dashboard view, users can place charts on a dashboard grid, similar to the current approach. We eliminate the live dashboard widgets and use the charts created in the data explorer as a basis for the live dashboard.
* Live dashboards can be configured with a refresh interval, and there is an option to override chart-specific time settings.
* We can migrate widgets that are currently used in the live dashboard to the data explorer (e.g., the traffic light component) so that no functionality gets lost.

As an implementation plan, I suggest to introduce this feature gradually over two releases. In the next release, we can keep the current live dashboard and introduce the new functionality to edit charts separately from the dashboard in the data explorer view. Users can then migrate their current dashboards to the new system. In the release thereafter, we can replace the live dashboard with the new data-explorer based dashboard.

I think this approach streamlines our development and will allow us to introduce more charts and more advanced chart settings in the future. I also think that this approach improves user experience since concepts are harmonized.

What do you think of this idea?

**Comment by tenthe** [comment]
Hi @dominikriemer,

Thanks for starting the discussion. I like the idea. However, I do not fully understand the migration strategy you suggested.

If I understand correctly, users will only be able to edit one widget in the data explorer. So, if we start with the data explorer in the next release, users will only be able to see one widget. Is this correct?

Alternatively, we could also introduce the new Data Explorer in the next release and leave the old Data Explorer and dashboard. And then remove the old Data Explorer and add the new dashboard in the following release.

**Reply by dominikriemer** [reply]
HI @tenthe 
for the next release, I propose to add support for auto-refreshing/live data to the data explorer dashboard and to add a feature to edit individual charts.
In this release, we can ask users to migrate existing live dashboards to the data explorer and in the release thereafter, we just move the dashboard functionality from the data explorer view to the dashboard view and remove the old dashboard view.

Here is an example:

![image](https://github.com/apache/streampipes/assets/33908476/cc71c8cd-c46d-4a01-930e-67891d97a0b5)

The `Data View` section will be moved to the `Dashboard` view in the release after the next release. The `Chart` section allows to edit charts, and the `Data View` section allows to place existing charts on a dashboard.

**Reply by tenthe** [reply]
Ok, thanks for the explanation. Now I understand what you mean.
I like the idea +1 from my side

**Comment by bossenti** [comment]
Thank you for starting this disscussion @dominikriemer 

I agree with you that having two data visualizazion functionalities can be confusing for the end user and increases the maintenance effort on our end. Therefore, I support your suggestion 👍🏼

**Comment by dominikriemer** [comment]
Hi,
I've just merged the data explorer changes into dev. There are now two views in the data explorer, one for creating dashboards and another for creating data views. Dashboards have a auto-refresh feature and are built from existing data views. Data views can also be viewed and edited independently from the dashboard. In release 0.98.0, we can move the dashboard functionality to the current dashboard view and keep the data views at the current location.

To fully cover the feature set of the existing live dashboard, we still need to add a few visualizations. To me, status light, traffic light and area chart seem to be the most important visualizations we currently don't have in the data explorer. Are there any other features which are supported by the current live dashboard that we should migrate?

Also, if you have any suggestions for improvement on the new features, add your comments here :-)
